_posts/2020/2020-02-11-unit-testing-in-react.md

```diff
@@ -0,0 +1,251 @@
+---
+layout: post
+title: Unit testing in React
+date: 2020-02-11
+header_image: public/code-coding-computer-cyberspace-270373.jpg
+header_overlay: true
+category: coding
+tags: ["javascript", "reactjs", "testing"]
+authors: ["Daniel"]
+about_authors: ["dhara"]
+---
+
+Automated Testing is one of the most important aspects of software development. You can test your code in different ways and levels, the most relevant
+of which are unit testing and integration testing.
+On the one hand, in unit testing, you test each one of your functions and components individually. On the other, in integration testing, you test
+how all those work along with each other to deliver all the fancy features you offer your users.
+
+As Adrià Fontcuberta pointed out at his remarkable talk at dotJS in Paris (check out this [blog post](/blog/events/dotjs-2019-in-paris-from-the-perspective-of-a-frontend-designer/){:target="_blank"} about the conference, as well as the [video](https://www.dotconferences.com/2019/12/adria-fontcuberta-the-pragmatic-front-end-tester){:target="_blank"} of the actual talk), tests give predictability and
+"help us sleep well at night".
+
+In this blog post, I'll show you useful hints for unit testing a simple component in React.
+
+## Getting started
+
+One of our most recent features at our storefront was a "Load more" button in the product pages. We could draft here a simplified version of this
+feature, and carry on to writing tests for it! This will help us to understand how tests work and how to ensure our tests are making our code
+reliable and predictable.
+
+We could start out with a simple component which has a button which can fetch (via an API) a fixed amount of, say, 1 product, and stores them
+in its state in order to subsequently display it. Our first shot could be:
+
+```javascript
+import React, { useState } from 'react';
+
+const Product = (product) => (
+  <div>
+    <div>{product.id}</div>
+    <div>{product.name}</div>
+    <div>{product.price}</div>
+  </div>
+)
+
+const fetchProducts = (page) => fetch(`http://localhost/products?page=${page}`)
+  .then(response => response.json())
+  .catch((error) => console.log(error))
+
+
+const ProductList = () => {
+  const [products, setProducts] = useState([])
+  const [page, setPage] = useState(1)
+  const loadProducts = async () => {
+    const response = await fetchProducts(page)
+    if(response) {
+      const newProducts = response.products
+      setProducts(products.concat(newProducts))
+      setPage(page + 1)
+    }
+  }
+  return <div>
+    <button onClick={loadProducts}>
+       Load more!
+    </button>
+    {products.map(product => <Product {...product} />)}
+  </div>
+}
+
+export default ProductList;
+```
+
+Before going on to how to test it, I must say I've prepared a sample repo for you to try out all the code in this post.
```

Comment:
Maybe you can write the exact word, instead of "it": "Before going on to how to test **it**, ..."

Reply:
I didn't understand this comment. What do you mean?